polyfill/lib/ecmascript.mjs

```diff
@@ -458,26 +544,77 @@ export const ES = ObjectAssign({}, ES2019, {
       const value = bag[property];
       if (value !== undefined) {
         any = any || {};
-        any[property] = ES.ToInteger(value);
+        if (property === 'calendar') {
+          // FIXME: this is terrible
+          const TemporalCalendar = GetIntrinsic('%Temporal.Calendar%');
+          any.calendar = TemporalCalendar.from(value);
+        } else if (property === 'era') {
+          any.era = value;
+        } else {
+          any[property] = ES.ToInteger(value);
+        }
       }
     }
     return any ? any : false;
   },
   ToRecord: (bag, fields) => {
     if (!bag || 'object' !== typeof bag) return false;
     const result = {};
-    for (const [property, defaultValue] of fields) {
+    for (const fieldRecord of fields) {
+      const [property, defaultValue] = fieldRecord;
       let value = bag[property];
       if (value === undefined) {
-        if (defaultValue === undefined) {
+        if (fieldRecord.length === 1) {
           throw new TypeError(`required property '${property}' missing or undefined`);
         }
         value = defaultValue;
       }
-      result[property] = ES.ToInteger(value);
+      if (property === 'calendar') {
+        // FIXME: this is terrible
+        const TemporalCalendar = GetIntrinsic('%Temporal.Calendar%');
+        result.calendar = TemporalCalendar.from(value);
+      } else if (property === 'era') {
+        result.era = value;
+      } else {
+        result[property] = ES.ToInteger(value);
+      }
     }
     return result;
   },
+  // field access in the following operations is intentionally alphabetical
+  ToTemporalDateRecord: (bag) => {
+    return ES.ToRecord(bag, [['day'], ['era', undefined], ['month'], ['year']]);
+  },
+  ToTemporalDateTimeRecord: (bag) => {
+    return ES.ToRecord(bag, [
+      ['day'],
+      ['era', undefined],
+      ['hour', 0],
+      ['microsecond', 0],
+      ['millisecond', 0],
+      ['minute', 0],
+      ['month'],
+      ['nanosecond', 0],
+      ['second', 0],
+      ['year']
+    ]);
+  },
+  ToTemporalMonthDayRecord: (bag) => {
+    return ES.ToRecord(bag, [['day'], ['month']]);
+  },
+  ToTemporalTimeRecord: (bag) => {
+    return ES.ToRecord(bag, [
+      ['hour', 0],
+      ['microsecond', 0],
+      ['millisecond', 0],
+      ['minute', 0],
+      ['nanosecond', 0],
+      ['second', 0]
+    ]);
+  },
+  ToTemporalYearMonthRecord: (bag) => {
+    return ES.ToRecord(bag, [['era', undefined], ['month'], ['year']]);
```

Comment:
Thanks for making this an abstract operation.  I wonder if this should be a method on the calendar, or maybe we should get the list of fields from the calendar and then *remove* 'day'.  This can be a follow-up issue.

Reply:
→ #666